data_structures/valid_parens.py

```diff
@@ -0,0 +1,25 @@
+class Solution:
+    def __init__(self):
+        self.stack = []
+
+    def isValid(self, s: str):
+        if not len(s):
+            return True
+        if len(s) % 2:
+            return False
+        for x in s:
+            if not len(self.stack):
+                self.stack.append(x)
+            else:
+                prev_idx = len(self.stack) - 1
+                if self.stack[prev_idx] == '(' and x == ')':
+                    del self.stack[prev_idx]
+                elif self.stack[prev_idx] == '[' and x == ']':
+                    del self.stack[prev_idx]
+                elif self.stack[prev_idx] == '{' and x == '}':
+                    del self.stack[prev_idx]
+                else:
```

Comment:
I think, if previous parentheses doesn't match current one - we can stop immediately, cause the string is invalid already.